src/main/java/org/opensearch/security/hasher/PasswordHasherFactory.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.hasher;
+
+import org.opensearch.common.settings.Settings;
+import org.opensearch.security.support.ConfigConstants;
+
+import static org.opensearch.security.support.ConfigConstants.BCRYPT;
+import static org.opensearch.security.support.ConfigConstants.PBKDF2;
+
+public class PasswordHasherFactory {
+
+    public static PasswordHasher createPasswordHasher(Settings settings) {
+
+        String algorithm = settings.get(
+            ConfigConstants.SECURITY_PASSWORD_HASHING_ALGORITHM,
+            ConfigConstants.SECURITY_PASSWORD_HASHING_ALGORITHM_DEFAULT
+        );
+
+        PasswordHasher passwordHasher;
+        switch (algorithm.toLowerCase()) {
+            case BCRYPT:
+                passwordHasher = getBCryptHasher(settings);
+                break;
+            case PBKDF2:
+                passwordHasher = getPBKDF2Hasher(settings);
+                break;
+            default:
+                throw new IllegalArgumentException("Password hashing algorithm not supported");
+        }
+        return passwordHasher;
+    }
+
+    private static PasswordHasher getBCryptHasher(Settings settings) {
+        int rounds = settings.getAsInt(
+            ConfigConstants.SECURITY_PASSWORD_HASHING_BCRYPT_ROUNDS,
+            ConfigConstants.SECURITY_PASSWORD_HASHING_BCRYPT_ROUNDS_DEFAULT
+        );
+        String minor = settings.get(
+            ConfigConstants.SECURITY_PASSWORD_HASHING_BCRYPT_MINOR,
+            ConfigConstants.SECURITY_PASSWORD_HASHING_BCRYPT_MINOR_DEFAULT
+        ).toUpperCase();
```

Comment:
i am not sure why we need a non-default value for this setting. the default should be always the standard, no?

Reply:
The default here (Y) is actually different to what Password4j itself defaults to (B) so it is consistent with the previous OpenBSDBCrypt generated hashes.

Honestly I am split between allowing this as an option or not. On one hand its not something that most users would want to change but on the other hand it might be good to have control over this if there is a new version of BCrypt in the future

Reply:
i think if something is not needed right away or not expected to change anytime soon, we should avoiding keeping it in config.

Reply:
Just to add more info: Y version of BCrypt is an older version than B so it still might be useful for customers to be able to change to a more up-to-date version if they wish. 

Another option is to make the default "B" and not make it configurable. That should work without issues for authentication as we validate with a BCrypt function based on the hash but my worry is that I am missing some other usages of hashing that might be affected by changing the default algorithm (like the replaceEnv functionality). 

Thoughts?

Reply:
> but my worry is that I am missing some other usages of hashing that might be affected by changing the default algorithm (like the replaceEnv functionality).

ok. in that case, let's keep ```Y```  as default, don't want to add regression / backward compatibility issues from such a change. 

I think we should open a separate PR if we want to make ```B``` as the new default cc @cwperks

Reply:
Changing the default is tricky because it would be a breaking change, right?

If a user wants to keep their existing internal user list and change the password hashing strategy then that would require updating the hashes of all internal users. One way to accomplish that is to keep the old hashes when starting a cluster after changing the hashing strategy. For each user that sends a request to the cluster, on the first request they would need to authenticate with the old hash and then in the backend we can compute the new hash and store it. On subsequent requests they would authenticate with the new hash.

Reply:
For existing hashes it should still work as when we validate we instantiate the hashing function based on the hash (the first two characters of the hash indicate if its 2y, 2b etc). But if we don't rehash every user after we change the default we will have the weird state where some stored hashes are of the 2y type and other newer hashes will be of 2b type.

Reply:
Agreed but I am not sure how I could approach this warning considering this configuration would be changed at deploy time (most likely). 

Is there an example of a similar warning for some other options that I can look at?

Reply:
> Is there an example of a similar warning for some other options that I can look at?

Not that I'm aware of. This may have to be added to the setting description itself. If we are planning to include this setting in the default opensearch.yml then a comment should be added to that yml as well IMO.